examples/api/lib/material/text_form_field/text_form_field.2.dart

```diff
@@ -0,0 +1,138 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+/// Flutter code sample for [TextFormField].
+
+const Duration kFakeHttpRequestDuration = Duration(seconds: 3);
+
+void main() => runApp(const TextFormFieldExampleApp());
+
+class TextFormFieldExampleApp extends StatelessWidget {
+  const TextFormFieldExampleApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return const MaterialApp(
+      home: TextFormFieldExample(),
+    );
+  }
+}
+
+class TextFormFieldExample extends StatefulWidget {
+  const TextFormFieldExample({super.key});
+
+  @override
+  State<TextFormFieldExample> createState() => _TextFormFieldExampleState();
+}
+
+class _TextFormFieldExampleState extends State<TextFormFieldExample> {
+  final TextEditingController controller = TextEditingController();
+  final GlobalKey<FormState> formKey = GlobalKey<FormState>();
+  String? forceErrorText;
+  bool isLoading = false;
+
+  @override
+  void dispose() {
+    controller.dispose();
+    super.dispose();
+  }
+
+  String? validator(String? value) {
+    if (value == null || value.isEmpty) {
+      return 'This field is required';
+    }
+    if (value.length != value.replaceAll(' ', '').length) {
+      return 'Username must not contain any spaces';
+    }
+    if (int.tryParse(value[0]) != null) {
+      return 'Username must not start with a number';
+    }
+    if (value.length <= 2) {
+      return 'Username should be at least 3 characters long';
+    }
+    return null;
+  }
+
+  void onChanged(String value) {
+    // Nullify forceErrorText if the input changed.
+    if (forceErrorText != null) {
+      setState(() {
+        forceErrorText = null;
+      });
+    }
+  }
+
+  Future<void> onSave() async {
+    // Providing a default value in case this was called on the
+    // first frame, the [fromKey.currentState] will be null.
+    final bool isValid = formKey.currentState?.validate() ?? false;
+    if (!isValid) {
+      return;
+    }
+
+    setState(() => isLoading = true);
+    final String? errorText = await validateUsernameFromServer(controller.text);
+
+    if (context.mounted) {
+      setState(() => isLoading = false);
+
+      if (errorText != null) {
+        setState(() {
+          forceErrorText = errorText;
+        });
+      }
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Material(
+      child: Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 24),
+        child: Center(
+          child: Form(
+            key: formKey,
+            child: Column(
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: <Widget>[
+                TextFormField(
+                  forceErrorText: forceErrorText,
+                  controller: controller,
+                  decoration: const InputDecoration(
+                    hintText: 'Please write a username',
```

Comment:
What can a user expect here if they also define a `errorText` for their `InputDecoration`? I tried this behavior out on master (explicitly defining an InputDecoration.errorText and using `validate` with AutovalidateMode set to `onUserInteraction`) and the behavior was the following:

1. On initial build the widget will show the `errorText` defined in `InputDecoration`.
2. When typing and validate returns a non-null string it replaces the `InputDecoration.errorText`.
3. When typing and validate returns a null string it shows the explicit `InputDecoration.errorText` that I defined.

I think with `forceErrorText` set and `InputDecoration.errorText` set, the result should be that the explicit `errorText` is never shown unless `forceErrorText` is set to `null`. This would be good to test.

Reply:
> I think with forceErrorText set and InputDecoration.errorText set, the result should be that the explicit errorText is never shown unless forceErrorText is set to null. This would be good to test.

I already added a test to cover this behavior [here](https://github.com/flutter/flutter/pull/132903/files#diff-d604873acef188d0441304f2f30ac15c648c1268a6752efee225c834ace7609fR1767)